### PR TITLE
Reduce Stress on Motor

### PR DESCRIPTION
**Pull request by brianhBR**
Improved stall detection, removed velocity based control, and adjusted braking and filter parameters to reduce motor stress.

### CODE REVIEW

**Review by patrickelectric — APPROVED** [review]
I don't know exactly what is happening here and what to expect/review,  but if it's working  ¯\_(ツ)_/¯
Maybe someone else that knows about the project should review it and approve.

**Review by Williangalvani — APPROVED** [review]
I am using it here and it seems to stall a lot less often and for shorter times, nice job!

**Comment by jaxxzer** [comment]
The compiled hex commit comes before the static float change. Should it come after?

File: Newton-Gripper/Newton-Gripper.ino [hunk 1]
```diff
@@ -170,29 +186,24 @@ void runSpeedFilter() {
     // Reject anything inside input deadzone
     if ( pw > INPUT_DZ ) {
     // Map valid PWM signals to (0.0 to 1.0]
```
Comment (jaxxzer) [inline Newton-Gripper/Newton-Gripper.ino]: this comment looks not to apply anymore

File: Newton-Gripper/Newton-Gripper.ino [hunk 2]
```diff
@@ -146,22 +166,18 @@ void loop() {
...
-  dir_t direction;
-  float voltage = readVoltage();
-  float maxduty = constrain(maxDuty(voltage),0.0f,MAX_DUTY_ABS);
-  float minduty = constrain(minDuty(voltage),0.0f,maxduty);
+  static float rawvelocity;
 
   // Save pulsein locally
   int16_t pulsewidth;
-  ATOMIC_BLOCK(ATOMIC_RESTORESTATE) {
+    ATOMIC_BLOCK(ATOMIC_RESTORESTATE) {
```
Comment (jaxxzer) [inline Newton-Gripper/Newton-Gripper.ino]: whitespace

File: Newton-Gripper/Newton-Gripper.ino [hunk 3]
```diff
@@ -146,22 +166,18 @@ void loop() {
...
     pulsewidth = pulsein;
   }
-
+
```
Comment (jaxxzer) [inline Newton-Gripper/Newton-Gripper.ino]: whitespace

File: Newton-Gripper/Newton-Gripper.ino [hunk 4]
```diff
@@ -170,29 +186,24 @@ void runSpeedFilter() {
...
-  } else {
-    // Stop motor if input is invalid
-    rawvelocity = 0.0f;
   }
 
   // Filter velocity
-  velocity = constrain(speedfilter.step(rawvelocity), -maxduty, maxduty);
-
+  velocity = constrain(speedfilter.step(rawvelocity), -1.0f, 1.0f);
+
```
Comment (jaxxzer) [inline Newton-Gripper/Newton-Gripper.ino]: whitespace

File: Newton-Gripper/Newton-Gripper.ino [hunk 5]
```diff
@@ -216,29 +227,38 @@ void runSpeedFilter() {
...
+ *
+ * Sets PWM output to motor controller
+ ******************************************************************************/
+void setMotorOutput(dir_t direction, dir_t limit, float velocity) {
+    
+    // Set output PWM timers
+  if ( direction == OPEN && limit != OPEN) {
     ATOMIC_BLOCK(ATOMIC_RESTORESTATE) {
-      OCR1A = abs(velocity)*MAX_COUNT;
+      OCR1A = fabs(velocity)*MAX_COUNT;
```
Comment (jaxxzer) [inline Newton-Gripper/Newton-Gripper.ino]: The change here might not be desirable (OCR1A is integer pwm register)

File: Newton-Gripper/Newton-Gripper.ino [hunk 6]
```diff
@@ -216,29 +227,38 @@ void runSpeedFilter() {
...
       OCR1B = 0;
     }
-  } else if ( direction == REVERSE && limit != REVERSE) {
+  } else if ( direction == CLOSE && limit != CLOSE) {
     ATOMIC_BLOCK(ATOMIC_RESTORESTATE) {
       OCR1A = 0;
-      OCR1B = abs(velocity)*MAX_COUNT;
-    }
+      OCR1B = fabs(velocity)*MAX_COUNT;
+    }
```
Comment (jaxxzer) [inline Newton-Gripper/Newton-Gripper.ino]: whitespace

File: Newton-Gripper/Newton-Gripper.ino [hunk 7]
```diff
@@ -263,29 +283,10 @@ float readVoltage() {
  *
  * Calculates stall current for given velocity and voltage (Amperes)
  ******************************************************************************/
-float stallCurrent(float velocity, float voltage) {
-  return (I_STALL_A0 + I_STALL_A1*voltage)*exp(I_STALL_B1*abs(velocity));
+float stallCurrent(float voltage) {
+  return constrain((voltage/(R_MOT*MOT_FS)),0,MOT_I_MAX);
```
Comment (jaxxzer) [inline Newton-Gripper/Newton-Gripper.ino:281]: can you please explain what is happening before/after and the reason for the change?

File: Newton-Gripper/Newton-Gripper.ino [hunk 8]
```diff
@@ -216,29 +227,38 @@ void runSpeedFilter() {
...
+      OCR1A = BRAKE*MAX_COUNT;
+      OCR1B = BRAKE*MAX_COUNT;
     }
   }
-
-  // Save last direction
-  lastdir = direction;
 }
 
-
```
Comment (jaxxzer) [inline Newton-Gripper/Newton-Gripper.ino:241]: whitespace

File: Newton-Gripper/Newton-Gripper.ino [hunk 9]
```diff
@@ -216,29 +227,38 @@ void runSpeedFilter() {
...
+
+/******************************************************************************
+ * void setMotorOutput()
+ *
+ * Sets PWM output to motor controller
+ ******************************************************************************/
+void setMotorOutput(dir_t direction, dir_t limit, float velocity) {
+    
+    // Set output PWM timers
+  if ( direction == OPEN && limit != OPEN) {
```
Comment (jaxxzer) [inline Newton-Gripper/Newton-Gripper.ino:238]: whitespace

File: Newton-Gripper/Newton-Gripper.ino [hunk 10]
```diff
@@ -216,29 +227,38 @@ void runSpeedFilter() {
...
     ATOMIC_BLOCK(ATOMIC_RESTORESTATE) {
-      OCR1A = abs(velocity)*MAX_COUNT;
+      OCR1A = fabs(velocity)*MAX_COUNT;
       OCR1B = 0;
     }
-  } else if ( direction == REVERSE && limit != REVERSE) {
+  } else if ( direction == CLOSE && limit != CLOSE) {
```
Comment (jaxxzer) [inline Newton-Gripper/Newton-Gripper.ino:243]: whitespace

File: Newton-Gripper/Newton-Gripper.ino (hunk 7 above)
Comment (brianhBR) [reply Newton-Gripper/Newton-Gripper.ino:281]: I removed the velocity from the stall current calculation to simplify the calculation. I think this was used to compensate for getting stuck on the end stops. The new calculation makes more sense physically by using the resistance of the motor to calculate the stall current at different voltages. The constrain function is used to set a max current below the calculated stall current at higher voltages.

File: Newton-Gripper/Newton-Gripper.ino [hunk 11]
```diff
@@ -167,32 +181,24 @@ void runSpeedFilter() {
...
-      // Map valid PWM signals to [-1.0 to 0.0)
-      rawvelocity = (float)(pw + INPUT_DZ)/(HALF_RANGE - INPUT_DZ)
-                    *(maxduty - minduty) - minduty;
+      rawvelocity = -1.0f;
     } else {
       // Stop motor if input is within input deadzone
       rawvelocity = 0.0f;
     }
-  } else {
-    // Stop motor if input is invalid
```
Comment (jaxxzer) [inline Newton-Gripper/Newton-Gripper.ino:184]: wouldn't you still want to stop the motor if you get an invalid signal?

Comment (brianhBR) [reply Newton-Gripper/Newton-Gripper.ino:184]: This was one of the root causes for the "beeping". Under heavy load (near stall) the PWM measurement is unreliable due to noise on the ground. We decided that it was better to continue with the current state between invalid signals. A better error rejection method is probably possible but this has been sufficient in our testing and we haven't gotten any negative feedback from beta testers.

**Comment by jaxxzer** [comment]
It works, so let's merge this. I think it could be better if we put more in to it, but perhaps now is not the time for that.

My two main concerns are:
- The speed filter immediately overrides the OCL braking action and may begin driving the motor again. This may result in high frequency oscillations between braking/driving the motor. perhaps the OCL action and the speed filter action should be decided at the same time (act on one or the other, instead of both)
- I think the pwm input signal should be filtered, and the motor should stop operating/reject commands when there is noise on the input. (for example require the last 5 successive readings to be valid for the input to be considered valid). This would add protection against high frequency changes in the output.

some other notes:
- The decision path is a bit convoluted and I would like to see it straightened out some day. For example, the 'limit' doesn't appear to need to be directional, it could just be a simpler 'over current' flag that would simplify some of the logic.
- i'm curious how well the current and voltage sampling/filtering works, these measurements usually need to be synchronized with the pwm to the motor (and with each other)
- it appears the current filter should be run before the speed filter, so that the speed filter is using the most recent output of the current filter
- it is strange to me that we continue to drive the motor when an invalid input signal is captured
- the floating point math seems excessive, particularly for calculating the delta-t to run the filters
- there are some potential memory leaks in the filter library, but it appears we won't encounter them with the way it is used (initialized once)
- the value assigned does not agree with the comments on this [line](https://github.com/bluerobotics/newton-firmware/pull/3/files#diff-b19f670d8db4454bed980a3e16356710R86)
- [This](https://github.com/bluerobotics/newton-firmware/pull/3/files#diff-8b2c81ae5e8c3032442d6aa8e5721aaaR195) constrain is redundant (because the filter saturation is set at 1.0)
- I'd like to have the build/deployment performed in CI for reproducibility

**Review by jaxxzer — APPROVED** [review]

**Pull request merged**